Sources/Hummingbird/Router/TransformingRouterGroup.swift

```diff
@@ -0,0 +1,59 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2021-2023 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import HTTPTypes
+import HummingbirdCore
+import NIOCore
+
+/// Internally used to transform RequestContext
+struct TransformingRouterGroup<InputContext: RequestContext, Context: RequestContext>: RouterMethods where Context.Source == InputContext {
+    typealias TransformContext = Context
+    let parent: any RouterMethods<InputContext>
+
+    struct ContextTransformingResponder: HTTPResponder {
+        typealias Context = InputContext
+        let responder: any HTTPResponder<TransformContext>
+
+        func respond(to request: Request, context: InputContext) async throws -> Response {
+            let newContext = TransformContext(source: context)
+            return try await self.responder.respond(to: request, context: newContext)
+        }
+    }
+
+    init(parent: any RouterMethods<InputContext>) {
+        self.parent = parent
+    }
+
+    /// Add middleware (Stub function as it isn't used)
+    @discardableResult func add(middleware: any RouterMiddleware<Context>) -> Self {
+        preconditionFailure("Cannot add middleware to context transform")
```

Comment:
I could add a separate protocol for Router types that can be a parent but don't hold middleware. To avoid adding the runtime error.

Reply:
I decided this isn't worth it. `TransformingRouterGroup` is a type the user will never get their hands on, so we can leave with the runtime error